virtcontainers/sandbox.go

```diff
@@ -1798,3 +1828,63 @@ func (s *Sandbox) AddDevice(info config.DeviceInfo) (api.Device, error) {
 
 	return b, nil
 }
+
+func (s *Sandbox) updateResources() error {
+	// the hypervisor.MemorySize is the amount of memory reserved for
+	// the VM and contaniners without memory limit
+
+	sumResources := specs.LinuxResources{
+		Memory: &specs.LinuxMemory{
+			Limit: new(int64),
+		},
+		CPU: &specs.LinuxCPU{
+			Period: new(uint64),
+			Quota:  new(int64),
+		},
+	}
+
+	for _, c := range s.config.Containers {
+		if m := c.Resources.Memory; m != nil && m.Limit != nil {
+			*sumResources.Memory.Limit += *m.Limit
+		}
+		if cpu := c.Resources.CPU; cpu != nil {
+			if cpu.Period != nil && cpu.Quota != nil {
+				*sumResources.CPU.Period += *cpu.Period
+				*sumResources.CPU.Quota += *cpu.Quota
+			}
+		}
+	}
+
+	sandboxVCPUs := uint32(utils.ConstraintsToVCPUs(*sumResources.CPU.Quota, *sumResources.CPU.Period))
```

Comment:
Use `utils.ConstraintsToVCPUs()` to calculate VCPUS is wrong. I have made the following test:
```bash
# 1. start a sandbox with default number of hypervisor cpu
# docker run -tid --runtime kata-runtime --annotation io.kubernetes.docker.type=podsandbox --network=none euleros bash
8589b666d0ff3d3211b3b3ebd2f8387e881d6d1068ca6518b1be6917e55a7208

# exec `lscpu` command in the container to view CPU numbers in the sandbox
 # docker exec -it 8589b666d0ff lscpu
Architecture:          x86_64 
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                1
On-line CPU(s) list:   0
Thread(s) per core:    1
Core(s) per socket:    1
Socket(s):             1
Vendor ID:             GenuineIntel

# 2. start a new container A with `--cpus 1.1` and join into the previous sandbox 
# docker run -tid --runtime kata-runtime --annotation io.kubernetes.docker.type=container --annotation io.kubernetes.sandbox.id=8589b666d0ff3d3211b3b3ebd2f8387e881d6d1068ca6518b1be6917e55a7208 --cpus 1.1 --network=none euleros bash
26984866a88783450b2d94cb0688db57fc939f4fd4d52416d68d30fad84ffd25

#  exec `lscpu` command in the container to view CPU numbers in the sandbox
 # docker exec -it 8589b666d0ff lscpu
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                3
On-line CPU(s) list:   0-2

# 3. start a new container B with `--cpus 0.8` and join into the previous sandbox
docker run -tid --runtime kata-runtime --annotation io.kubernetes.docker.type=container --annotation io.kubernetes.sandbox.id=8589b666d0ff3d3211b3b3ebd2f8387e881d6d1068ca6518b1be6917e55a7208 --cpus 0.8 --network=none euleros bash
fdd5c1a10429ee77b8984911061ff62c6d479dfc131da30ea0193e9a614c0166

#  exec `lscpu` command in the container to view CPU numbers in the sandbox
# docker exec -it fdd5 lscpu
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                2
On-line CPU(s) list:   0,1
Thread(s) per core:    1
Core(s) per socket:    1
Socket(s):             2
Vendor ID:             GenuineIntel
```
After  I create container B in step 3, I get the CPU numbers is **2**, however expected value is **3**，I think we should propose a new method to calculate VCPU numbers.

Reply:
Hey thanks for the the comment nice catch! We should add this as a test and agree need to fix ConstraintsToVCPUs as needed.

Nice way to test pods in docker  I am am trying the annotation flag  but not worked for me what docker version are you using?

Reply:
@jcvenegas 
Hah this is a internal version we are using。
You can patch docker with https://github.com/moby/moby/pull/37289.